pkg/controller/nodeobservatory/controller.go

```diff
@@ -0,0 +1,379 @@
+
+	if err == nil {
+		// Forget about the #AddRateLimited history of the key on every successful synchronization.
+		// This ensures that future processing of updates for this key is not delayed because of
+		// an outdated error history.
+
+		if requeue == false {
+			n.queue.Forget(key)
+		} else {
+			// Requeue requested
+			n.queue.AddAfter(key, BaseDelay)
+		}
+
+		return true
+	}
+
+	if n.queue.NumRequeues(key) < 5 {
+		// Re-enqueue the key rate limited. Based on the rate limiter on the
+		// queue and the re-enqueue history, the key will be processed later again.
+		n.queue.AddRateLimited(key)
+		return true
+	}
+
+	// Retries exceeded. Forgetting for this reconciliation loop
+	n.queue.Forget(key)
+	return true
+}
+
+func (n *NodeObservatory) reconcile(kluster *v1.Kluster) (bool, error) {
+
+	n.cleanUpInformers()
+
+	if err := n.createAndWatchNodeInformerForKluster(kluster); err != nil {
+		return true, err
+	}
+
+	return false, nil
+}
+
+func (n *NodeObservatory) cleanUpInformers() {
+	n.nodeInformerMap.Range(
+		func(key, value interface{}) bool {
+			if _, exists, _ := n.Factories.Kubernikus.Kubernikus().V1().Klusters().Informer().GetIndexer().GetByKey(key.(string)); !exists {
+				if i, ok := n.nodeInformerMap.Load(key); ok {
+					informer := i.(NodeInformer)
+					informer.close()
+				}
+				n.nodeInformerMap.Delete(key)
+			}
+			return true
+		},
+	)
+}
+
+func (n *NodeObservatory) createAndWatchNodeInformerForKluster(kluster *v1.Kluster) error {
+	key, err := cache.MetaNamespaceKeyFunc(kluster)
+	if err != nil {
+		return err
+	}
+
+	_, exists := n.nodeInformerMap.Load(key)
+	if !exists {
+		nodeInformer, err := newNodeInformerForKluster(n.Clients.Satellites, kluster, make(chan struct{}))
+		if err != nil {
+			return err
+		}
+
+		nodeInformer.SharedIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if kluster, err = n.getKlusterByKey(key); err != nil {
+					n.logger.Log(err)
+					return
+				}
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, addHandler := range n.addEventHandlers {
+					addHandler(kluster, obj.(*api_v1.Node))
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				if kluster, err = n.getKlusterByKey(key); err != nil {
+					n.logger.Log(err)
+					return
+				}
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, updateHandler := range n.updateEventHandlers {
+					updateHandler(kluster, oldObj.(*api_v1.Node), newObj.(*api_v1.Node))
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if kluster, err = n.getKlusterByKey(key); err != nil {
+					n.logger.Log(err)
+					return
+				}
+				n.handlersMux.RLock()
+				defer n.handlersMux.RUnlock()
+				for _, deleteHandler := range n.deleteEventHandlers {
+					deleteHandler(kluster, obj.(*api_v1.Node))
+				}
+			},
+		})
+
+		n.nodeInformerMap.Store(
+			key,
+			nodeInformer,
+		)
+
+		cache.WaitForCacheSync(
```

Comment:
Isn't this called to early? `informer.run()` is called down below. Looks like a deadlock to me.

Reply:
Not sure we even need this here. `WaitForCacheSync` is commonly used wait before starting a `workqueue` but we don't have a work queue for precessing the node events.

Reply:
I thought it's also common to wait for an initial sync before processing. But true, so far I've only seen it in combination with workqueues too. Removed.